### DISCUSSION TITLE
Thanks for creating this! Having a bit of trouble getting Sixel output working.

### DISCUSSION THREAD

**Discussion by danschwarz** [GuardKenzie/chafa.py#15] [Q&A] [answered]
Thanks for creating this! I was hoping someone would write Python bindings for Chafa; you've done that and more!

I've been trying to integrate this with my branch of enhancements to the Toot urwid app. (Source [here](https://github.com/danschwarz/toot/tree/sixel3).  I am building urwid widgets that support sixel and ANSI image rendering. Chafa's ANSI rendering would be an improvement on what I have, and the Kitty + iTerm image support is an added bonus.

I have a test program based on your example code that is capable of loading a PNG and displaying it in ANSI block characters.  Sixel output isn't working yet.  Your docs indicate that this may be related to a lack of Termdb info for my terminal emulator.  I'm using Xterm (372) on Ubuntu 22.04; 

TermInfo.detect_capabilties returns:
CanvasMode: CanvasMode.CHAFA_CANVAS_MODE_FGBG
PixelMode:  PixelMode.CHAFA_PIXEL_MODE_SYMBOLS

printing with (fallback=True) doesn't help.

Chafa binary with -f sixel does output sixel images properly to this terminal emulator.

Any suggestions?

**Comment by GuardKenzie** [comment, marked as answer]
Hihi! Thanks for taking an interest in this project! I'm glad you like it! :heart:

Are you setting your canvas config's pixel mode mode to `CHAFA_PIXEL_MODE_SIXELS` or are you setting them according to the output of `TermInfo.detect_capabilities()`?

The `TermInfo.detect_capabilities()` will try it's best to detect the capabilities of your terminal based on your environment variables (mainly `$TERM`) and xterm does not seem to set those differently when launching it with `-ti vt340` as opposed to just regular ol' xterm. Thus chafa.py will resort to the mode it will definitely be able to display.

**Reply by danschwarz** [reply]
I have tried setting `config.pixel_mode = CHAFA_PIXEL_MODE_SIXELS` and this results in no output.

According to Nick Black, it's possible to detect Sixel support by querying the terminal with escape codes ([Send Device Attributes](https://nick-black.com/dankwiki/index.php/Sixel#Detecting_support)). If you can do this, it may be more reliable than examining $TERM. 

My Xterm reports

`❯ printf '\e[c'`
`~/dev ❯ 63;1;2;4;6;9;15;16;22;28c `

the output contains a `4` so the terminal is indicating Sixel support.

**Reply by GuardKenzie** [reply]
Thanks for pointing this out! I had a read through [this document](https://www.x.org/docs/xterm/ctlseqs.pdf) and decided to try to implement this. It was not trivial though since emitting this control sequence puts the device attributes on `STDIN` which is hard to read from python without executing an return of some sort. I think I figured out a pretty good solution using some C code to read the attributes and binding that using ctypes in chafa.py but it still needs some polishing before release. 

Stay tuned!

**Reply by danschwarz** [reply]
Not sure if this will help the specific situation you're facing, but @AnonymouX47 has Python code to query terminal info in the term-image project,  for example in [utils.py](https://github.com/AnonymouX47/term-image/blob/main/src/term_image/utils.py) see the get_terminal_size method in that file for an example. Probably could adapt that to execute the Send Device Attributes query.

**Reply by GuardKenzie** [reply]
Thanks for that! I took a look and it seems to be pretty similar to what I ended up implementing (see commit 4c0ec7f). This will be implemented in the next minor release which will hopefully be pretty soon (I'm having trouble finding willpower to write the docs for it :P)

**Comment by danschwarz** [comment]
Update; 

Forcing `CHAFA_PIXEL_MODE_SIXELS` is now working. Aspect ratio is wrong but that's easily corrected.

![image](https://user-images.githubusercontent.com/3261094/215303855-156d632f-6e95-4177-90b0-f203c075c775.png)
